xarray/core/variable.py

```diff
@@ -422,62 +422,40 @@ def _broadcast_indexes(self, key):
-            # more than one (unlabelled) arrays
-            if len([v for k, v in indexes.items()
-                    if not isinstance(v, (integer_types, slice))]) > 1:
-                raise IndexError("Indexing with multiple unlabelled arrays "
-                                 "is not allowed.")
-            # multi-dimensional unlabelled array
-            if any([v.ndim > 1 for k, v in indexes.items()
-                    if not isinstance(v, integer_types)]):
-                raise IndexError("Indexing with a multi-dimensional unlabelled"
-                                 "array is not allowed.")
-            # convert the array into Variable
-            for k, v in indexes.items():
-                if not hasattr(v, 'dims'):
-                    indexes[k] = type(self)([k], v)
-                    example_v = v
-
-        for k, v in indexes.items():
-            # Found unlabelled array
-            if not isinstance(v, (Variable, integer_types, slice)):
-                if (v.ndim > example_v.ndim or
-                        any([example_v.ndim != v.ndim for k, v
-                             in indexes.items() if isinstance(v, Variable)])):
-                    raise IndexError("Broadcasting failed because dimensions "
-                                     "does not match.")
-                else:
-                    _, indexes[k], _ = _broadcast_compat_data(example_v, v)
-
-        index_tuple = tuple(indexes.get(d, slice(None)) for d in self.dims)
-        index_tuple = indexing.expanded_indexer(index_tuple, self.ndim)
-
-        # comput dims
-        dims = []
-        for i, d in enumerate(self.dims):
-            if d in indexes.keys():
-                if isinstance(v, Variable):
-                    for d in v.dims:
-                        if d not in dims:
-                            dims.append(d)
-            else:
-                dims.append(d)
+        key = self._item_key_to_tuple(key)  # key is a tuple
+        # key is a tuple of full size
+        key = indexing.expanded_indexer(key, self.ndim)
+        basic_indexing_types = integer_types + (slice,)
+        if all([isinstance(k, basic_indexing_types) for k in key]):
+            return self._broadcast_indexes_basic(key)
+        else:
+            return self._broadcast_indexes_advanced(key)
+
+    def _broadcast_indexes_basic(self, key):
+        dims = tuple(dim for k, dim in zip(key, self.dims)
+                     if not isinstance(k, integer_types))
+        return dims, key
+
+    def _broadcast_indexes_advanced(self, key):
+        variables = []
 
-        return dims, index_tuple
+        for dim, value in zip(self.dims, key):
+            if isinstance(value, slice):
+                value = np.arange(self.sizes[dim])[value]
+
+            try:  # TODO we need our own Exception.
+                variable = as_variable(value, name=dim)
+            except ValueError as e:
+                if "cannot set variable" in str(e):
```

Comment:
I agree. Can you switch this for `as_variable`?

Inherit from `ValueError` for backwards compatibility. Something like `MissingDimensionsError` should be more broadly useful for cases like this where we can't safely guess a dimension name.